File: nodejs/src/ws/stomp.js
'use strict';

const { WebSocketServer } = require('ws');
const { generate, VALID_SIZES_KB } = require('../payload/generator');

const STOMP_NULL = '\x00';

function parseStompFrame(data) {
  data = data.replace(/\x00$/, '');
  const parts = data.split('\n\n');
  const head = parts[0] || '';
  const body = parts[1] || '';
  const lines = head.split('\n');
  const command = (lines[0] || '').trim();
  const headers = {};
  for (let i = 1; i < lines.length; i++) {
    const idx = lines[i].indexOf(':');
    if (idx > 0) headers[lines[i].slice(0, idx).trim()] = lines[i].slice(idx + 1).trim();
  }
  return { command, headers, body };
}

function buildStompFrame(command, headers = {}, body = '') {
  const lines = [command];
  for (const [k, v] of Object.entries(headers)) lines.push(`${k}:${v}`);
  lines.push('', body);
  return lines.join('\n') + STOMP_NULL;
}

function sockjsSend(ws, message) {
  ws.send('a' + JSON.stringify([message]));
}

function handleStompFrame(ws, frame, subscriptions) {
  const { command, headers, body } = parseStompFrame(frame);

  switch (command) {
    case 'CONNECT':
      sockjsSend(ws, buildStompFrame('CONNECTED', { version: '1.2', 'heart-beat': '0,0' }));
      break;
    case 'SUBSCRIBE': {
      const subId = headers.id || 'sub-0';
      subscriptions[subId] = headers.destination;
      if (headers.receipt) {
        sockjsSend(ws, buildStompFrame('RECEIPT', { 'receipt-id': headers.receipt }));
      }
      break;
    }
    case 'SEND': {
      const dest = headers.destination || '';
      if (dest.endsWith('/mock') || dest === '/app/mock') {
        let req = {};
        if (body) {
          try {
            req = JSON.parse(body);
          } catch {
            req = {};
          }
        }
        const size = parseInt(req.size, 10) || 1;
        const format = req.format || 'json';
        const method = req.method || 'GET';
        const resource = req.resource || 'items';
        let out;
        try {
          out = generate(size, format, method, resource);
        } catch (err) {
          out = JSON.stringify({ error: err.message, validSizes: VALID_SIZES_KB });
        }
        for (const [subId, subDest] of Object.entries(subscriptions)) {
          if (subDest.includes('/topic/response') || subDest.endsWith('/response')) {
            sockjsSend(
              ws,
              buildStompFrame('MESSAGE', {
                subscription: subId,
                'message-id': `msg-${Math.floor(Math.random() * 999999)}`,
                destination: subDest,
              }, out)
            );
          }
        }
      }
      break;
    }
    default:
      break;
  }
}

function attachStompWs(server) {
  const wss = new WebSocketServer({ noServer: true });

  server.on('upgrade', (request, socket, head) => {
    const url = new URL(request.url, 'http://localhost');
    if (!url.pathname.endsWith('/websocket')) return;

    wss.handleUpgrade(request, socket, head, (ws) => {
      wss.emit('connection', ws, request);
    });
  });

  wss.on('connection', (ws) => {
    const subscriptions = {};
    ws.send('o');

    ws.on('message', (raw) => {
      let messages;
      try {
        messages = JSON.parse(raw.toString());
      } catch {
        return;
      }
      if (!Array.isArray(messages)) messages = [messages];
      for (const msg of messages) {
        if (typeof msg === 'string') handleStompFrame(ws, msg, subscriptions);
      }
    });
  });

  return wss;
}

function sockjsInfo(req, res) {
  res.json({
    websocket: true,
    cookie_needed: false,
    origins: ['*:*'],
    entropy: Math.floor(Math.random() * 9_000_000_000) + 1_000_000_000,
  });
}

module.exports = { attachStompWs, sockjsInfo };

File: go/internal/ws/stomp.go
package ws

import (
	"encoding/json"
	"math/rand"
	"net/http"
	"strconv"
	"strings"

	"github.com/SidathWeerasinghe/mock-services/go/internal/payload"
	"github.com/gorilla/websocket"
)

const stompNull = "\x00"

func SockJSInfo(w http.ResponseWriter, r *http.Request) {
	w.Header().Set("Content-Type", "application/json")
	_ = json.NewEncoder(w).Encode(map[string]interface{}{
		"websocket":     true,
		"cookie_needed": false,
		"origins":       []string{"*:*"},
		"entropy":       rand.Int63n(9_000_000_000) + 1_000_000_000,
	})
}

func StompWebSocket(w http.ResponseWriter, r *http.Request) {
	conn, err := Upgrader.Upgrade(w, r, nil)
	if err != nil {
		return
	}
	defer conn.Close()

	subs := make(map[string]string)
	_ = conn.WriteMessage(websocket.TextMessage, []byte("o"))

	for {
		_, raw, err := conn.ReadMessage()
		if err != nil {
			break
		}
		var messages []string
		if err := json.Unmarshal(raw, &messages); err != nil {
			continue
		}
		for _, frame := range messages {
			handleStompFrame(conn, frame, subs)
		}
	}
}

func handleStompFrame(conn *websocket.Conn, frame string, subs map[string]string) {
	cmd, headers, body := parseStompFrame(frame)
	switch cmd {
	case "CONNECT":
		sockJSSend(conn, buildStompFrame("CONNECTED", map[string]string{
			"version":    "1.2",
			"heart-beat": "0,0",
		}, ""))
	case "SUBSCRIBE":
		subID := headers["id"]
		if subID == "" {
			subID = "sub-0"
		}
		subs[subID] = headers["destination"]
		if receipt := headers["receipt"]; receipt != "" {
			sockJSSend(conn, buildStompFrame("RECEIPT", map[string]string{"receipt-id": receipt}, ""))
		}
	case "SEND":
		dest := headers["destination"]
		if strings.HasSuffix(dest, "/mock") || dest == "/app/mock" {
			var req struct {
				Size     interface{} `json:"size"`
				Format   string      `json:"format"`
				Method   string      `json:"method"`
				Resource string      `json:"resource"`
			}
			if body != "" {
				_ = json.Unmarshal([]byte(body), &req)
			}
			size := toInt(req.Size, 1)
			format := req.Format
			if format == "" {
				format = "json"
			}
			method := req.Method
			if method == "" {
				method = "GET"
			}
			resource := req.Resource
			if resource == "" {
				resource = "items"
			}
			out, err := payload.Generate(size, format, method, resource)
			if err != nil {
				b, _ := json.Marshal(map[string]interface{}{
					"error": err.Error(), "validSizes": payload.ValidSizesKB,
				})
				out = string(b)
			}
			for subID, subDest := range subs {
				if strings.Contains(subDest, "/topic/response") || strings.HasSuffix(subDest, "/response") {
					sockJSSend(conn, buildStompFrame("MESSAGE", map[string]string{
						"subscription": subID,
						"message-id":   "msg-" + strconv.Itoa(rand.Intn(999999)),
						"destination":  subDest,
					}, out))
				}
			}
		}
	}
}

func parseStompFrame(data string) (command string, headers map[string]string, body string) {
	data = strings.TrimSuffix(data, stompNull)
	parts := strings.SplitN(data, "\n\n", 2)
	head := parts[0]
	if len(parts) > 1 {
		body = parts[1]
	}
	lines := strings.Split(head, "\n")
	command = strings.TrimSpace(lines[0])
	headers = make(map[string]string)
	for _, line := range lines[1:] {
		if i := strings.Index(line, ":"); i > 0 {
			headers[strings.TrimSpace(line[:i])] = strings.TrimSpace(line[i+1:])
		}
	}
	return command, headers, body
}

func buildStompFrame(command string, headers map[string]string, body string) string {
	var b strings.Builder
	b.WriteString(command)
	for k, v := range headers {
		b.WriteString("\n")
		b.WriteString(k)
		b.WriteString(":")
		b.WriteString(v)
	}
	b.WriteString("\n\n")
	if body != "" {
		b.WriteString(body)
	}
	b.WriteString(stompNull)
	return b.String()
}

func sockJSSend(conn *websocket.Conn, message string) {
	frame := "a" + string(mustJSON([]string{message}))
	_ = conn.WriteMessage(websocket.TextMessage, []byte(frame))
}

func mustJSON(v interface{}) []byte {
	b, _ := json.Marshal(v)
	return b
}
